lib/logger.js

```diff
@@ -0,0 +1,653 @@
+    this.indexMeta = false
+    this.app = 'default'
+    this.env = undefined
+    this.baseBackoffMs = constants.BASE_BACKOFF_MILLIS
+    this.maxBackoffMs = constants.MAX_BACKOFF_MILLIS
+
+    // Initialize internal instance variables
+    this[kLineLengthTotal] = 0
+    this[kBuffer] = []
+    this[kMeta] = {}
+    this[kIsLoggingBackedOff] = false
+    this[kAttempts] = 0
+    this[kRequestDefaults] = undefined
+    this[kFlusher] = null
+    this[kReadyToSend] = []
+    this[kIsSending] = false
+    this[kTotalLinesReady] = 0
+    this[kBackoffMs] = constants.BASE_BACKOFF_MILLIS
+
+    let useHttps = true
+    let withCredentials = false
+    let tags = undefined
+    let timeout = constants.DEFAULT_REQUEST_TIMEOUT
+    let hostname = os.hostname()
+    let mac = undefined
+    let ip = undefined
+
+    if (has(options, 'level')) {
+      const level = options.level.toUpperCase()
+      if (!constants.LOG_LEVELS.includes(level)) {
+        const err = new Error('Invalid level')
+        err.meta = {
+          got: level
+        , expectedOneOf: constants.LOG_LEVELS
+        }
+        throw err
+      }
+      this.level = level
+    }
+
+    if (has(options, 'tags')) {
+      tags = options.tags
+      if (typeof tags === 'string') {
+        tags = tags.split(constants.TAGS_RE)
+      }
+      if (!Array.isArray(tags)) {
+        const err = new TypeError('tags should be passed as a String or an Array')
+        err.meta = {got: options.tags}
+        throw err
+      }
+      tags = tags
+        .filter((tag) => { return tag !== '' })
+        .map((tag) => { return tag.trim() })
+        .join(',')
+    }
+
+    if (has(options, 'meta')) {
+      const meta = options.meta
+      if (typeOf(meta) !== 'object') {
+        const err = new TypeError('meta needs to be an object of key-value pairs')
+        err.meta = {got: meta}
+        throw err
+      }
+      this[kMeta] = {...meta}
+    }
+
+    if (has(options, 'timeout')) {
+      if (!Number.isInteger(options.timeout)) {
+        const err = new TypeError('timeout must be an Integer')
+        err.meta = {got: options.timeout}
+        throw err
+      }
+      if (options.timeout > constants.MAX_REQUEST_TIMEOUT) {
+        throw new Error(
+          `timeout cannot be longer than ${constants.MAX_REQUEST_TIMEOUT}`
+        )
+      }
+      timeout = options.timeout
+    }
+
+    if (has(options, 'hostname')) {
+      checkStringParam(options.hostname, 'Hostname')
+      if (!constants.HOSTNAME_CHECK.test(options.hostname)) {
+        throw new Error('Invalid hostname')
+      }
+      hostname = options.hostname
+    }
+
+    if (has(options, 'mac')) {
+      checkStringParam(options.mac, 'MAC Address')
+      if (!constants.MAC_ADDR_CHECK.test(options.mac)) {
+        throw new Error('Invalid MAC Address format')
+      }
+      mac = options.mac
+    }
+
+    if (has(options, 'ip')) {
+      checkStringParam(options.ip, 'IP Address')
+      if (!isIP(options.ip)) {
```

Comment:
Do you need `checkStringParam` if you're using `isIP`?

Reply:
`checkStringParam` validates the string type (and specifically throws a TypeError) and `isIP` validates the format

Reply:
Hm okay, I thought `isIP()` would effectively restrict to string types, but on a quick test it does work with a Buffer object which is not(?) what we want? Maybe that warrants another question: should we handle Buffers and strings equally? This convention seems somewhat consistent in Node's standard library (and here with net's `isIP()`)

Reply:
`checkStringParam` is a nice way to ensure there's a value to test.  Otherwise, falsey values could fail when they shouldn't. You lost me at Buffer, though.  Why would instantiation input ever be a buffer for this sort of parameter?

Reply:
@jakedipity gonna merge this for now, but let's circle back on these last points so I can fully understand your thoughts.  Thanks for thinking it through!